src/coffee/sync/product-sync.coffee

```diff
@@ -36,23 +36,105 @@ class ProductSync extends BaseSync
+
+    return {
+      variantUpdateActions
+      addVariantActions
+      removeVariantActions
+      changeMasterVariantAction
+    }
+
+  _doMapActions: (diff, newObj, oldObj) ->
     # Update actions needs to be sorted and executed in particular order.
-    variantActions = @_mapActionOrNot 'variants', => @_utils.actionsMapVariants(diff, old_obj, new_obj)
-
-    allActions = []
-    # First redundant variants need to be removed to prevent SKU conflicts, e.g. in case of chaging the master variant
-    allActions.push variantActions.filter (action) -> action.action is 'removeVariant'
-    # Before adding new variants, all attributes needs to be updated to the newest value to prevent adding a new variant with different
-    # sameForAll attribute value than the variants in the CTP product.
-    allActions.push @_mapActionOrNot 'attributes', => @_utils.actionsMapAttributes(diff, old_obj, new_obj, @sameForAllAttributeNames)
-    allActions.push variantActions.filter (action) -> action.action is 'addVariant'
-    allActions.push @_mapActionOrNot 'base', => @_utils.actionsMapBase(diff, old_obj)
-    allActions.push @_mapActionOrNot 'references', => @_utils.actionsMapReferences(diff, old_obj, new_obj)
-    allActions.push @_mapActionOrNot 'prices', => @_utils.actionsMapPrices(diff, old_obj, new_obj)
-    allActions.push @_mapActionOrNot 'images', => @_utils.actionsMapImages(diff, old_obj, new_obj)
+    allActions =  []
+
+    # base product data update actions
+    allActions.push @_mapActionOrNot 'base', => @_utils.actionsMapBase(diff, oldObj)
+    allActions.push @_mapActionOrNot 'references', => @_utils.actionsMapReferences(diff, newObj, oldObj)
     allActions.push @_mapActionOrNot 'categories', => @_utils.actionsMapCategories(diff)
-    allActions.push @_mapActionOrNot 'categoryOrderHints', => @_utils.actionsMapCategoryOrderHints(diff, old_obj)
-    _.flatten allActions
+    allActions.push @_mapActionOrNot 'categoryOrderHints', => @_utils.actionsMapCategoryOrderHints(diff, oldObj)
+
+    # generate variant related update actions
+    variantActionGroups = @_doMapVariantActions(newObj, oldObj)
+
+    # apply remove/add/changeMaster variant in the following order:
+    # - remove all variants except master variant
+    # - apply setAttribute/image .. update actions
+    # - add new variants
+    # - promote a new masterVariant
+    # - remove masterVariant
+
+    # divide removeVariant actions to two groups - one for masterVariant, second for normal variants
+    partitionedRemoveActionsByMasterVariant = _.partition variantActionGroups.removeVariantActions, (action) =>
+      oldObj.masterVariant and @_utils.matchesByIdOrKeyOrSku(action, oldObj.masterVariant)
+
+    allActions.push @_mapActionOrNot 'variants', => partitionedRemoveActionsByMasterVariant[1] # normal variants
+    allActions.push variantActionGroups.variantUpdateActions # already white/black listed from @_diffVariant
```

Comment:
Based on the discussion from the last week, we should first update existing variants. Because if you have some `sameForAll` attribute, we should first update existing variants before we try to add new variants with this new value - otherwise we would run into an error because existing variants would still have the old value